File: windows/src/Slipstream.Core/Discovery/CachedEndpointStrategy.cs
using Slipstream.Core.Net;

namespace Slipstream.Core.Discovery;

/// <summary>
/// Spec §5 S1. The common case: both target networks are stable, so the last known
/// address is usually still correct and answers in about 50 ms.
/// </summary>
public sealed class CachedEndpointStrategy(EndpointCache cache, PeerProbe probe) : IDiscoveryStrategy
{
    public string Name => "cached-endpoint";

    public async Task<DiscoveredPeer?> FindAsync(LocalNetwork network, CancellationToken cancellationToken)
    {
        var endpoint = cache.Get(network.Key);
        if (endpoint is null) return null;

        // Re-validate on read: the cache is on-disk state, not a trusted source.
        if (!LanGuard.IsLocal(endpoint.Address)) return null;

        return await probe(endpoint, cancellationToken);
    }
}

File: android/core/src/main/kotlin/com/slipstream/core/discovery/CachedEndpointStrategy.kt
package com.slipstream.core.discovery

import com.slipstream.core.net.LocalNetwork
import java.net.InetAddress
import java.net.InetSocketAddress

/**
 * S1: direct-connect to the last-known endpoint for the current network, if any. Spec §5:
 * the common case, expected to resolve in tens of milliseconds.
 */
class CachedEndpointStrategy(
    private val cache: EndpointCache,
    private val probe: PeerProbe,
) : DiscoveryStrategy {

    override val name: String = "cached-endpoint"

    override suspend fun find(network: LocalNetwork): DiscoveredPeer? {
        val cached = cache.get(network.key) ?: return null
        val endpoint = parseEndpoint(cached) ?: return null
        return probe.probe(endpoint)
    }

    private fun parseEndpoint(value: String): InetSocketAddress? {
        val separator = value.lastIndexOf(':')
        if (separator <= 0 || separator == value.length - 1) return null

        val host = value.substring(0, separator)
        val port = value.substring(separator + 1).toIntOrNull() ?: return null
        if (port !in 1..65535) return null

        return try {
            InetSocketAddress(InetAddress.getByName(host), port)
        } catch (e: java.net.UnknownHostException) {
            null
        }
    }
}
